### ISSUE TITLE
Inspect network returns invalid gateway (CIDR notation) if none was set during creation

### ISSUE THREAD

**Issue by shadjiiski** [docker/docker#26522]
**Description**
When using the CLI or the remote API, to `create` a network only the **name** property is required and everything else is set to a default value if no other was set. However, it looks like the gateway in the IPAM configuration defaults to invalid value (or at least an invalid value is returned from a successive `inspect` command) - Returned addresses are in CIDR notation, for example `172.20.0.1/16` rather than `172.20.0.1`. This makes the output invalid.

**Steps to reproduce the issue:**
- Reproduce with CLI
  1. `docker network create gateway-test-2`
  2. inspect the network
   ```ts
docker network inspect gateway-test-2
[
    {
        "Name": "gateway-test-2",
        "Id": "01120199318e7c798f2d4aefe627e26fbb8bb217704ee86a34aaf98434ae5c8c",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.28.0.0/16",
                    "Gateway": "172.28.0.1/16"
                }
            ]
        },
        "Internal": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
   ```
- Reproduce with Remote API
  1. Create a network
  ```ts
curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" -X POST -d '{"Name":"gateway-test-1"}' http:/networks/create
{"Id":"aecef3e535340631e4fb60223dae32e7f2858940333f8aa8549826ad7d633e2a","Warning":""}
  ```
  2. Inspect the network
  ```ts
 curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" http:/networks/aecef3e535340631e4fb60223dae32e7f2858940333f8aa8549826ad7d633e2a | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   316  100   316    0     0     37      0  0:00:08  0:00:08 --:--:--    77
{
    "Containers": {},
    "Driver": "bridge",
    "EnableIPv6": false,
    "IPAM": {
        "Config": [
            {
                "Gateway": "172.27.0.1/16",
                "Subnet": "172.27.0.0/16"
            }
        ],
        "Driver": "default",
        "Options": null
    },
    "Id": "aecef3e535340631e4fb60223dae32e7f2858940333f8aa8549826ad7d633e2a",
    "Internal": false,
    "Labels": {},
    "Name": "gateway-test-1",
    "Options": {},
    "Scope": "local"
}
  ```

**Describe the results you received:**
The gateway that docker defaulted to are invalid:
```ts
172.27.0.1/16
172.28.0.1/16
```

**Describe the results you expected:**
Expected  valid gateway addresses:
```ts
172.27.0.1
172.28.0.1
```

**Additional information you deem important (e.g. issue happens only occasionally):**
Looks like only the output is invalid. I have attached an alpine container to one of the test networks
```
docker run -itd --name alp10 --network gateway-test-1 alpine sh
winpty docker attach alp10
```
and it is properly configured:
```
/ # route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.27.0.1      0.0.0.0         UG    0      0        0 eth0
172.27.0.0      0.0.0.0         255.255.0.0     U     0      0        0 eth0
```

**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 23:54:00 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 28
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.17-boot2docker
Operating System: Boot2Docker 1.12.1 (TCL 7.2); HEAD : ef7d0b4 - Thu Aug 18 21:18:06 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9 MiB
Name: test-host-1
ID: VW3L:L572:64HH:TSKM:ZQCU:7VAQ:X6TI:A25A:IRAZ:IH37:BESI:YJ4R
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 31
 Goroutines: 51
 System Time: 2016-09-13T08:53:44.189326685Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
 com.docker.network.driver.overlay.bind_interface=eth0
 provider=virtualbox
Cluster Store: etcd://192.168.99.50:4001
Cluster Advertise: 192.168.99.100:2376
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
I am testing this on Win 8.1 with `docker-machine.exe version 0.8.0, build b85aac1` and VirtualBox Version 5.0.20 r106931 (reported from UI)

**Comment by justincormack** [comment]
Not sure I would say it is invalid. CIDRs are correct and useful. If there is an inconsistency though that is a problem.

**Comment by shadjiiski** [comment]
@justincormack, as far as I know (not an expert on networking), CIDRs are meant to represent a mask or a range and I am not really sure what `172.22.0.1/16` means in the context of a gateway address. I read this as "the gateway can be found on address 172.22.0.1, but the .0.1 part may change". This is the reason why I used the word 'invalid'.

Also, from my initial post you can clearly see that the network configuration of the container reports a gateway with a standard IPv4 (and not in CIDR notation), so yes, there is a bit of inconsistency here. What is more, docker actually is not allowing to set the gateway in CIDR notation:
```ts
$ docker network create --subnet 172.50.0.0/16 --gateway 172.50.0.1/16 gateway-test
Error response from daemon: invalid gateway address 172.50.0.1/16 in Ipam configuration
```
whereas the standard IPv4 version is allowed:
```ts
$ docker network create --subnet 172.50.0.0/16 --gateway 172.50.0.1 gateway-test
7127911ba24ec682595f4f7d57ead08e2d50dc4efde70a1ae9dd7953f446de7f
```
Furthermore, if I set the gateway manually, the inspect command reports the address with no CIDR notation:
```ts
$ docker network inspect gateway-test
[
    {
        "Name": "gateway-test",
        "Id": "7127911ba24ec682595f4f7d57ead08e2d50dc4efde70a1ae9dd7953f446de7f",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.50.0.0/16",
                    "Gateway": "172.50.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
```
This is another reason for me to think the CIDR notation (however useful for masks and ranges) is invalid in the context of a gateway address.

**Comment by justincormack** [comment]
cc @aboch

**Comment by aboch** [comment]
I agree the network gateway should not be reported in CIDR notation.

While I can hit the issue on 1.12.1, I am not able to reproduce the problem on master, neither via cli nor via remote api.

Looks like this has already been fixed on master.

Will look for which docker/docker PR took care of this.

**Comment by aboch** [comment]
It looks like it was fixed by #25564

This issue can be closed.

**Comment by shadjiiski** [comment]
@aboch, thanks for the update!
